Lesson003: Web server graceful shutdown (#9)
contributing-to-kubernetes/go-examples: lesson-003-web-server-graceful-shutdown/README.md

# Lesson 3: Web server graceful shutdown

Taking as a foundation the webserver created in [lesson 000](../lesson-000-web-server/), we will be adding `graceful shutdown` functionality to it.

In order to achieve this, we will be using `channels` and `goroutines`

## Knowledge bits

### Goroutines

A `goroutine` is a lightweight thread of execution managed by the Go runtime; it is the way we have to run a piece of code concurrently with the original calling code, as we can see explained in the [oficial documentation.](https://tour.golang.org/concurrency/1)

A small example could be:

```
func say(s string) {
	for i := 0; i < 5; i++ {
		time.Sleep(100 * time.Millisecond)
		fmt.Println(s)
	}
}

func main() {
	go say("k8s")
	say("contribute")
}
```

> Calling `go printMyString()` is where we start a new `goroutine`, calling our already defined `printMyString` function, while the `main` function runs in its own `goroutine` (called the _main goroutine_)

Reviewer: Need to update the name of the function from `printMyString` to `say`.

Author: 🤦

### Channels

Channels are a concurrency synchronization technique we can use in Go.

We can create a channel using the keyword `chan`, and we can transport data of only one data type.

From the [official documention](https://tour.golang.org/concurrency/2) we can see how data can be sent and received from the channel.

```
ch := make(chan int)

ch <- v    // Send v to channel ch.
v := <-ch  // Receive from ch, and
           // assign value to v.
```

**_Unbuffered_ vs _buffered_ channels**

Channels are created **blocking by default** (or _unbuffered_)

> Unbuffered channels are those which both goroutines (one sending data to the buffer, another waiting to receive the resource) require to be ready to exchange some data. If a goroutine is trying to send a resource to an unbuffered channel and there is no goroutine waiting to receive that resource, the channel will lock the sender and make it wait.

Reviewer: ```suggestion
If we try to send a resource to an unbuffered channel, the channel will lock (preventing us from sending anything else into the channel) until someone else reads from it.
This essentially means that an unbuffered channel is restricted to never having more than 1 element inside of it.
```

Reviewer: Small rewrite here.
Suggesting to remove go routines from the wording since normal functions can also send and receive from channels.

And viceversa.

On the other hand, buffered channels have capacity and they are able to keep a number of resources.

> The only times buffered channels will lock goroutines are: when a sender tries to send a resource and the channel is full or when a goroutine tries to get a resource and the channel is empty.

Reviewer: ```suggestion
The only times buffered channels will lock goroutines are when a sender tries to send a resource and the channel is full or when a goroutine tries to get a resource and the channel is empty.
```

We can use this in our benefit and block our server from being closed until all pending requests have been served when we receive a specific signal.

### Signals

Signals are software interrupts sent to a program to indicate that an event has happened.

In this example we will take care of two specific signals: `os.Interrupt` and `syscall.SIGTERM`:

- `os.Interrupt` this is tipically the signal sent when we type Control-C, which normally causes the program to exit.
- `syscall.SIGTERM` is usually sent when you want to give the process an opportunity to clean up before termination.

## Lesson content

After getting a quick intro to `channels` and `goroutines`, we will dive into our lesson.

As we initially stated, the current implementation is based on our server from [lesson 000](../lesson-000-web-server/).

To achieve _graceful shutdown_ we added the following changes:

> - created a `sigint` channel; we will use it to notify our goroutine we have received a signal: `os.Interrupt` or `syscall.SIGTERM` in this case

Reviewer: ```suggestion
- created a `sigint` channel; we will use it to notify our goroutine we have received a signal: `os.Interrupt` or `syscall.SIGTERM`
```

> - with `server.SetKeepAlivesEnabled(false)` we set the server to not keep alive any connection (which in fact is the desired effect of having a gracefull shutdown behavior)

Reviewer: Add something along the lines of: the desired behaviour of a graceful shutdown is to allow the server to finish processing any requests already received before the app received a termination signal.
The first step is to disable "keep alive" tcp connections (https://godoc.org/net/http#Server.SetKeepAlivesEnabled) before proceeding with the graceful shutdown of the server https://godoc.org/net/http#Server.Shutdown

> - create the `done` channel; this one will be used to let the main goroutine we have finished the graceful shutdown.

Reviewer: useful to mention that adding this in main allows us to run code only after the app is shut down - if needed.

In order to achieve this, we will be using `channels` and `goroutines`

## Knowledge bits

### Channels

Channels are a concurrency synchronization technique we can use in Go.

We can create a channel using the keyword `chan`, and we can transport data of only one data type.

From the [official documention](https://tour.golang.org/concurrency/2) we can see how data can be sent and received from the channel.

```
ch := make(chan int)

ch <- v    // Send v to channel ch.
v := <-ch  // Receive from ch, and
           // assign value to v.
```

In our case we want to use channels along with `goroutines`.

### Goroutines

A `goroutine` is a lightweight thread of execution managed by the Go runtime; it is the way we have to run a piece of code concurrently with the original calling code, as we can see explained in the [oficial documentation.](https://tour.golang.org/concurrency/1)

A small example could be:

```
func printMyString() {
    fmt.Println("my string")
}

func main() {
    go printMyString()
    time.Sleep(100 * time.Millisecond)
    fmt.Println("printMyString function")
}
```

> In line 6 is where we start a new `goroutine`, calling our already defined `printMyString` function, while the `main` function runs in its own `goroutine` (called the _main goroutine_)

Reviewer: re: `In line 6`
I think it'd be good to avoid referencing line numbers unless we explicitly add them :smile:

Reviewer: Also maybe look for another example: this one doesn't make it too clear what we are doing with concurrency - it could very well be that we run `printMyString` outside of a go routine and we would see the same.

**_But how are we going to use both features together?_**

We need to point something about channels: they are **blocking by default**.

> We can use this in our benefit and block our server from being closed until all pending requests have been served when we receive a specific signal.

Reviewer: I think this may need some bit more of explanation: buffered v. unbuffered channels.
I don't think that someone who hasn't heard of a (un)buffered channel will understand the idea here.

Author: Sure! Will add a more useful explanation 👍

## Lesson content

After getting a quick intro to `channels` and `goroutines`, we will dive into our lesson.

> As we initially stated, the current implementation is based on our server from lesson0.

Reviewer: Another link to lesson 000 may be helpful here.

> To achieve _graceful shutdown_ we added the following main changes:

Reviewer: nit:
```suggestion
To achieve _graceful shutdown_ we added the following changes:
```

> - created a `sigint` channel; we will use it to notify our goroutine we have received a signal: `os.Interrupt` or `syscall.SIGTERM` in this case

Reviewer: It may be useful to add another "knowledge bits" section to talk about signals.

Author: Makes sense 👍
